pkg/zeek/addr.go

```diff
@@ -24,68 +25,85 @@ func (t *TypeOfAddr) String() string {
 	return "addr"
 }
 
-func (t *TypeOfAddr) Parse(value []byte) (net.IP, error) {
+func EncodeAddr(a net.IP) zval.Encoding {
+	ip := a.To4()
+	if ip == nil {
+		ip = net.IP(a)
+	}
+	return []byte(ip)
+}
+
+func DecodeAddr(value []byte) (net.IP, error) {
 	if value == nil {
 		return nil, ErrUnset
 	}
-	return UnsafeParseAddr(value)
+	switch len(value) {
+	case 4, 16:
+		return net.IP(value), nil
+	}
+	return nil, errors.New("failure trying to decode IP address that is not 4 or 16 bytes long")
 }
 
-func (t *TypeOfAddr) Format(value []byte) (interface{}, error) {
-	return t.Parse(value)
+func (t *TypeOfAddr) Parse(in []byte) (zval.Encoding, error) {
+	ip, err := UnsafeParseAddr(in)
+	if err != nil {
+		return nil, err
+	}
+	return EncodeAddr(ip), nil
 }
 
-func (t *TypeOfAddr) New(value []byte) (Value, error) {
-	if value == nil {
+func (t *TypeOfAddr) New(zv zval.Encoding) (Value, error) {
+	if zv == nil {
 		return &Unset{}, nil
 	}
-	ip, err := t.Parse(value)
+	ip, err := DecodeAddr(zv)
 	if err != nil {
 		return nil, err
 	}
-	return &Addr{Native: ip}, nil
+	return NewAddr(ip), nil
 }
 
-type Addr struct {
-	Native net.IP
+type Addr net.IP
+
+func NewAddr(a net.IP) *Addr {
+	p := Addr(a)
+	return &p
 }
 
-func (a *Addr) String() string {
-	return a.Native.String()
+func (a Addr) String() string {
+	return (net.IP)(a).String()
 }
 
-func (a *Addr) Type() Type {
+func (a Addr) Type() Type {
 	return TypeAddr
 }
 
-func (a *Addr) Encode(dst zval.Encoding) zval.Encoding {
-	b := []byte(a.Native.String())
-	return zval.AppendValue(dst, b)
+func (a Addr) Encode(dst zval.Encoding) zval.Encoding {
+	return zval.AppendValue(dst, EncodeAddr(net.IP(a)))
 }
 
 // Comparison returns a Predicate that compares typed byte slices that must
 // be TypeAddr with the value's address using a comparison based on op.
 // Only equality operands are allowed.
-func (a *Addr) Comparison(op string) (Predicate, error) {
+func (a Addr) Comparison(op string) (Predicate, error) {
 	compare, ok := compareAddr[op]
 	if !ok {
 		return nil, fmt.Errorf("unknown addr comparator: %s", op)
 	}
-	pattern := a.Native
+	pattern := net.IP(a)
 	return func(e TypedEncoding) bool {
-		typeAddr, ok := e.Type.(*TypeOfAddr)
-		if !ok {
+		if _, ok := e.Type.(*TypeOfAddr); !ok {
```

Comment:
Nit: I think `e.Type != TypeAddr` would be clearer.

Reply:
ok